core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
@lindong28 : My thoughts are the following. (1) When reacting to a disk failure, it would be useful to handle this in order, e.g., removing the partition from ReplicaManager, followed by removing the log from LogManager. If you only remove the log from LogManager w/o removing the partition from ReplicaManager, then a request may hit an unexpected exception when trying to access the log, which is not ideal. (2) The notification to ZK could take a long time if ZK is not performing. Doing the ZK notification in a background thread reduces the potential latency impact to the client request. (3) We may implement a more sophisticated disk failure detection module (e.g., proactively verifying CRCs in the log) in the future. Having a DiskFailureChannel allows us to integrate such a module easier (such a module only needs to interact directly with DiskFailureChannel, instead of ReplicaManager).

I agree that adding a separate disk failure handling thread adds a bit complexity, but probably not too much. To avoid building up the queue, perhaps DiskFailureChannel can just maintain a disk dir set instead of a queue.